File: CS/MapperCore/Mapper.cs
using System;
using System.Linq.Expressions;
using MapperExpression.Core;
using MapperExpression.Exceptions;
using System.Reflection;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics.Contracts;

namespace MapperExpression
{

    /// <summary>
    ///Base class for each access to mapper
    /// </summary>
    public static class Mapper
    {
        #region Variables


        private static Func<Type, object> constructorFunc;

        #endregion

        #region Public methods

        /// <summary>
        /// Maps the specified source.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TTarget">The type of the dest.</typeparam>
        /// <param name="source">the source object.</param>
        /// <param name="name">The name.</param>
        /// <returns>
        /// A new  object of <typeparamref name="TTarget"></typeparamref>
        /// </returns>
        public static TTarget Map<TSource, TTarget>(TSource source, string name = null)
            where TSource : class
            where TTarget : class
        {
            // No source, no target.
            if (source == null)
                return null;
            TTarget result = null;
            try
            {
                MapperConfiguration<TSource, TTarget> mapper = GetMapper<TSource, TTarget>(name);
                Func<TSource, TTarget> query = mapper.GetFuncDelegate();
                if (query != null)
                {
                    result = query(source);
                    // Action performed after the mapping
                    mapper.ExecuteAfterActions(source, result);
                }
            }
            catch (System.Exception)
            {
                throw;
            }
            return result;
        }

        /// <summary>
        /// Maps the specified source to the target.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TTarget">The type of the target.</typeparam>
        /// <param name="source">The source.</param>
        /// <param name="target">The target.</param>
        /// <param name="name">The name.</param>
        public static void Map<TSource, TTarget>(TSource source, TTarget target, string name = null)
         where TSource : class
         where TTarget : class
        {

            Contract.Requires(source != null);
            Contract.Requires(target != null);
            TTarget result = null;
            try
            {
                MapperConfiguration<TSource, TTarget> mapper = GetMapper<TSource, TTarget>(name);
                Action<TSource, TTarget> query = mapper.GetDelegateForExistingTarget() as Action<TSource, TTarget>;
                if (query != null)
                {
                    query(source, target);
                    // Action performed after the mapping
                    mapper.ExecuteAfterActions(source, result);
                }
            }
            catch (System.Exception)
            {
                throw;
            }
        }

        /// <summary>
        /// Gets the query expression.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TTarget">The type of the dest.</typeparam>
        /// <returns></returns>
        public static Expression<Func<TSource, TTarget>> GetQueryExpression<TSource, TTarget>()
            where TSource : class
            where TTarget : class
        {
            return GetMapper<TSource, TTarget>().GetLambdaExpression();
        }

        /// <summary>
        /// Creates a mapper.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TTarget">The type of the dest.</typeparam>
        /// <returns></returns>
        public static MapperConfiguration<TSource, TTarget> CreateMap<TSource, TTarget>(string name = null)
            where TSource : class
            where TTarget : class
        {
            // We do not use the method because it GetMapper throw an exception if not found
            MapperConfigurationBase map = MapperConfigurationCollectionContainer.Instance.Find(typeof(TSource), typeof(TTarget), name);
            if (map == null)
            {
                string finalName = string.IsNullOrEmpty(name) ? "s" + MapperConfigurationCollectionContainer.Instance.Count.ToString() : name;
                map = new MapperConfiguration<TSource, TTarget>(finalName);
                MapperConfigurationCollectionContainer.Instance.Add(map);
            }
            return map as MapperConfiguration<TSource, TTarget>;
        }
        /// <summary>
        /// Indicates the injection service used
        /// </summary>
        /// <param name="constructor">The constructor.</param>
        public static void ConstructServicesUsing(Func<Type, object> constructor)
        {
            constructorFunc = constructor;
        }

        /// <summary>
        /// Remove all mappers
        /// </summary>
        ///<remarks>Use for your units test only</remarks>
        public static void Reset()
        {
            MapperConfigurationCollectionContainer.Instance.Clear();
        }
        /// <summary>
        /// Gets the mapper.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <param name="name">The name.</param>
        /// <returns></returns>
        public static MapperConfiguration<TSource, TDest> GetMapper<TSource, TDest>(string name = null)
            where TSource : class
            where TDest : class
        {

            return (GetMapper(typeof(TSource), typeof(TDest), name) as MapperConfiguration<TSource, TDest>);

        }
        /// <summary>
        /// Initialise the mappers.
        /// </summary>
        /// <remarks>Use only the application initialization</remarks>
        public static void Initialize()
        {
            MapperConfigurationCollectionContainer configRegister = MapperConfigurationCollectionContainer.Instance;
            for (int i = 0; i < configRegister.Count; i++)
            {
                configRegister[i].Initialize(constructorFunc);
            }


        }

        /// <summary>
        /// Gets the delegate of the mapping.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <returns></returns>
        public static Func<TSource, TDest> GetQuery<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            return GetMapper<TSource, TDest>().GetFuncDelegate();
        }
        /// <summary>
        /// Gets the properties not mapped.
        /// </summary>
        public static PropertiesNotMapped GetPropertiesNotMapped<TSource, TDest>(string name = null)
             where TSource : class
            where TDest : class
        {
            var mapper = GetMapper<TSource, TDest>(name);
            return mapper.GetPropertiesNotMapped();
        }
        #endregion

        #region Internals methods        


        internal static MapperConfigurationBase GetMapper(Type tSource, Type tDest, string name = null)
        {
            MapperConfigurationBase mapConfig = MapperConfigurationCollectionContainer.Instance.Find(tSource, tDest, name);
            if (mapConfig == null)
                throw new NoFoundMapperException(tSource, tDest);

            return mapConfig;
        }

        #endregion
    }
    /// <summary>
    ///Base class for each access to mapper(simplified access)
    /// </summary>
    /// <typeparam name="TTarget">The type of the dest.</typeparam>
    public static class Mapper<TTarget>
         where TTarget : class
    { 
        /// <summary>
        /// Maps the specified source.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <param name="source">The source.</param>
        /// <param name="name">The name.</param>
        public static TTarget Map<TSource>(TSource source, string name = null)
            where TSource : class
        {

            return Mapper.Map<TSource, TTarget>(source, name);
        }
    }
}

File: Mapper/Mapper.vb

Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Linq.Expressions

Imports System.Data.Linq
Imports MapperExpression.Core



''' <summary>
''' Class de base pour l'accès au mapping
''' </summary>
Public Class Mapper

    Private Shared current As MapperConfigurationBase

#Region "Méthodes publiques"
    ''' <summary>
    ''' Maps the specified source.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <param name="source">The business.</param>
    ''' <returns></returns>
    Public Shared Function Map(Of TSource As Class, TDest As Class)(source As TSource) As TDest
        Try

            Dim query = GetMapper(Of TSource, TDest)().GetDelegate()
            If query IsNot Nothing Then
                Return query(source)
            End If
        Catch ex As Exception

            Throw ex
        End Try
        Return Nothing
    End Function


    ''' <summary>
    ''' Gets the query expression.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetQueryExpression(Of TSource As Class, TDest As Class)() As Expression(Of Func(Of TSource, TDest))
        Return GetMapper(Of TSource, TDest)().GetLambdaExpression()
    End Function
    ''' <summary>
    ''' Creates the map.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function CreateMap(Of TSource As Class, TDest As Class)() As MapperConfiguration(Of TSource, TDest)

        Dim map = MapperConfigurationRegister.Instance.Find(GetType(TSource), GetType(TDest))
        If map Is Nothing Then
            map = New MapperConfiguration(Of TSource, TDest)()
            MapperConfigurationRegister.Instance.Add(map)
        End If
        Return TryCast(map, MapperConfiguration(Of TSource, TDest))
    End Function
    ''' <summary>
    ''' Efface tout les mappeur existants
    ''' </summary>
    Public Shared Sub Reset()
        MapperConfigurationRegister.Instance.Clear()
    End Sub
    ''' <summary>
    ''' Gets the data load options for linqToSql.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetDataLoadOptionsLinq(Of TSource As Class, TDest As Class)() As DataLoadOptions
        Dim map = GetMapper(Of TSource, TDest)()
        Return map.GetDataLoadOptionsLinq()
    End Function
    ''' <summary>
    ''' Initialises les mappeurs.
    ''' </summary>
    Public Shared Sub Initialize()
        Dim configRegister = MapperConfigurationRegister.Instance
        For Each mapper As MapperConfigurationBase In configRegister
            mapper.CreateMappingExpression()
        Next
    End Sub
    ''' <summary>
    ''' Gets the query.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetQuery(Of TSource As Class, TDest As Class)() As Func(Of TSource, TDest)
        Return GetMapper(Of TSource, TDest)().GetDelegate()
    End Function
#End Region

#Region "Méthodes privées"

    Private Shared Function GetMapper(Of TSource As Class, TDest As Class)() As MapperConfiguration(Of TSource, TDest)
        If current Is Nothing OrElse (current IsNot Nothing AndAlso current.TypeDest.FullName <> GetType(TDest).FullName AndAlso current.TypeSource.FullName <> GetType(TSource).FullName) Then
            current = GetMapper(GetType(TSource), GetType(TDest))
        End If

        Return TryCast(current, MapperConfiguration(Of TSource, TDest))

    End Function

    Private Shared Function GetMapper(tSource As Type, tDest As Type) As MapperConfigurationBase
        Dim mapConfig = MapperConfigurationRegister.Instance.Find(tSource, tDest)
        If mapConfig IsNot Nothing Then
            Return mapConfig
        Else
            Throw New Exception("Impossible de trouver la configuration entre le type source" + tSource.Name + " et le type destination" + tDest.Name)
        End If

    End Function
#End Region
End Class

